### ISSUE TITLE
Explicitly support building with -fno-code

### ISSUE THREAD

**Issue by nh2** [haskell/cabal#1176]
**Summary by @ezyang.** There is no way to ask Setup to just typecheck a package. It sort of works if you pass `-fno-code` via `--ghc-options` but that's purely accidental.

---

I use `cabal build --ghc-options="-Wall -fno-code"` to quickly get all errors/warnings from a project and show them in an editor.

However, when I have something like this

```
executable Shared.so
  hs-source-dirs:
    apps
  main-is:
    Shared.hs
  build-depends:
      base >= 4 && <= 5
  ghc-options:
    -optl-shared -optc-DMODULE=Shared -no-hs-main -fPIC -shared -dynamic
  cc-options: -DMODULE=Shared -shared
  ld-options: -shared /home/niklas/opt/haskell-7.4/lib/ghc-7.4.2/libHSrts-ghc7.4.2.so
```

in my cabal file that only builds a shared object, I get:

```
Warning: the following files would be used as linker inputs, but linking is not being done: dist/build/...
ghc: no input files
Usage: For basic information, try the `--help' option.
```

Should we get this warning although we explicitly said `-fno-code`?

If yes, can we add an option to disable this kind of warning so that it does not clutter the warnings I am actually interested in (to show in the editor)?

---

**Update:**

(The original title was _Add option to disable non-linking warnings with -fno-code_.)

I just realized that there is an actual bug in this:

After the warning, due to the ghc _error_ cabal will just **stop** and build no further. This should not happen as there is no code to build!

**Comment by AndrewBrinker** [comment]
This issue seems to persist, and is currently causing minor frustration in using SublimeHaskell, whose default build mechanism uses two builds: first a standard `cabal build` and then `cabal build -v0 --ghc-options="-fforce-recomp -Wall -fno-code"` to gather warnings. This issue is causing SublimeHaskell to display a bothersome error pane every time a project is recompiled (which happens every save). It's a minor issue, but a fix would be appreciated.

**Comment by nh2** [comment]
A tip from @rwbarton just now on `#ghc`, we could use `-e` which is like `ghci`, but terminates.

TH works in ghci (due to use of bytecode), so this should be almost as fast as `-fno-code` (it still generates the bytecode which is _some_ code, but in my experience that's much faster than even `-O0`).

**Comment by mightybyte** [comment]
I'd like to add another vote for this ticket.  I'm not sure what the UI should be (`cabal typecheck` or `cabal build --typecheck` come to mind).  Whether it's called "typecheck", "no-code", "fast", etc is not a huge issue (although "no-code" never seemed very intuitive to me personally).  The important thing in my mind is to have this as an explicit cabal option somewhere so the fastest compile loop is more discoverable.

**Comment by sboosali** [comment]
the `-e` option doesn't work for me, because I'm using `cabal build` as a replacement for `ghci` in a project where the interpreter can't load at all because of linking errors (i.e. and thus cabal repl, ghcid, dante, etc can't help).

**Comment by ice1000** [comment]
> I'd like to add another vote for this ticket. I'm not sure what the UI should be (`cabal typecheck` or `cabal build --typecheck` come to mind). Whether it's called "typecheck", "no-code", "fast", etc is not a huge issue (although "no-code" never seemed very intuitive to me personally). The important thing in my mind is to have this as an explicit cabal option somewhere so the fastest compile loop is more discoverable.

I like this idea! Any update for this?

**Comment by sboosali** [comment]
same. having a subcommand is better for discovery too.

e.g. someone who tab-completes `cabal t` for testing will notice
`typecheck` and they start using it.

also, even if though it's (currently) equivalent to just a few options,
it's one less thing a user needs to do, and they can trust that it's the
right way to do it.

**Comment by andreasabel** [comment]
In v1-cabal, this works for just type-checking:
```
cabal v1-build --builddir=dist-no-code \
	  --ghc-options=-fno-code \
	  --ghc-options=-fwrite-interface
```
In v2-cabal, it seems that passing new `--ghc-options` will cause ALL THE DEPENDENCIES to be rebuilt.  That defeats the purpose of quick type-checking.

Example: We build ordinarily:
```
$ cabal v2-build -w ghc-8.0.2
Resolving dependencies...
Build profile: -w ghc-8.0.2 -O1
In order, the following will be built (use -v for more details):
 - haskell-lexer-1.1 (lib:haskell-lexer) (requires build)
 - prettyprinter-1.7.0 (lib) (requires build)
 - pretty-show-1.10 (lib) (requires build)
 - prettyprinter-ansi-terminal-1.1.2 (lib) (requires build)
 - BNFC3-3.0 (lib) (configuration changed)
 - BNFC3-3.0 (exe:bnfc3) (configuration changed)
...
```
After that, we want to just type-check:
```
$ cabal v2-build -w ghc-8.0.2 --ghc-options=-fno-code --ghc-options=-fwrite-interface
Resolving dependencies...
Build profile: -w ghc-8.0.2 -O1
In order, the following will be built (use -v for more details):
 - base-orphans-0.8.4 (lib) (requires build)
 - clock-0.8.2 (lib) (requires build)
 - colour-2.3.5 (lib) (requires build)
 - containers-0.6.4.1 (lib) (requires build)
 - haskell-lexer-1.1 (lib:haskell-lexer) (requires build)
 - alex-3.2.6 (exe:alex) (requires build)
 - microlens-0.4.12.0 (lib) (requires build)
 - mtl-2.2.2 (lib) (requires build)
 - stm-2.5.0.0 (lib) (requires build)
 - string-qq-0.0.4 (lib) (requires build)
 - transformers-compat-0.6.6 (lib) (requires build)
 - extra-1.7.9 (lib) (requires build)
 - ansi-terminal-0.11 (lib) (requires build)
 - th-abstraction-0.4.2.0 (lib) (requires build)
 - binary-0.8.8.0 (lib) (requires build)
 - happy-1.20.0 (exe:happy) (requires build)
 - transformers-base-0.4.5.2 (lib) (requires build)
 - microlens-mtl-0.2.0.1 (lib) (requires build)
 - ansi-wl-pprint-0.6.9 (lib) (requires build)
 - microlens-th-0.4.3.9 (lib) (requires build)
 - text-1.2.4.1 (lib) (requires build)
 - monad-control-1.0.2.3 (lib:monad-control) (requires build)
 - optparse-applicative-0.16.1.0 (lib) (requires build)
 - prettyprinter-1.7.0 (lib) (requires build)
 - pretty-show-1.10 (lib) (requires build)
 - prettyprinter-ansi-terminal-1.1.2 (lib) (requires build)
 - BNFC3-3.0 (lib) (configuration changed)
 - BNFC3-3.0 (exe:bnfc3) (configuration changed)
``` 
`^C` is the only answer to such madness!

**Comment by fgaz** [comment]
@andreasabel I think the issue you mention is #3579 (ghc-options is not available in [per-package options](https://cabal.readthedocs.io/en/3.4/cabal-project.html#package-configuration-options))

**Comment by jneira** [comment]
The command toonly typecheck suggested by @andreasabel:
 `cabal v2-build --ghc-options=-fno-code --ghc-options=-fwrite-interface`

is more useful after #7973, cause now `ghc-options` applies only to all local packages. So change it does not trigger the recompilation of dependencies (aka remote packages).

@nh2 would be it enough for your use case? how could be it improved to fit it? create a new command would be difficult to get done

**Comment by kindaro** [comment]
This is a long awaited improvement. I built Cabal and tried it out today. It works for a single component. It does not work for `cabal build all`. It seems that Cabal tries to link the dependent components _(such as internal libraries)_, but there is nothing to link because nothing was built!

**Comment by kindaro** [comment]
If adding a new command `cabal typecheck` is still on the table, I say it should behave like this:

* Accept the same options as `cabal build`.
* Produce the same standard output and standard error as `cabal build`.
* Produce no object files.
* Run faster.

Even though thanks to `haskell-language-server` one needs to call Cabal only rarely, it is still a good sanity check for when one is about to commit some code, for continuous integration or whenever `haskell-language-server` gets stuck or cannot start.

**Comment by andreasabel** [comment]
Yes, it should be like `-O0`, in particular, use its own build directory.
The main annoyance with the current workaround (a separate `cabal.project` file with suitable `ghc-options`) is that `builddir` isn't accepted in `cabal.project`, and has to be passed on the command line manually all the time, like in https://github.com/agda/agda/blob/a47f8762978e878ecf6ed6081fa3772acc3499d1/Makefile#L238 .
In contrast `-O0` stores build artefacts in its own `.../noopt/...` folders so that the regular build artefacts (`-O1`) are not overwritten.

**Comment by ulysses4ever** [comment]
@andreasabel sorry for a basic question, but why do you need to store `-no-code` artifacts separately? For -O0, a separate directory is needed so that you don't rebuild every time you switch configuration, but that shouldn't apply to `-no-code` as it shouldn't, in theory, generate any different artifacts?

**Comment by andreasabel** [comment]
> it shouldn't, in theory, generate any different artifacts?

The typechecking-only build tree contains the cabal configuration, interface files, and even some object files (e.g. `Paths_*.o`, `setup/Main.o`, modules using `TemplateHaskell`).  Until evidence to the contrary is produced, I must assume there are differences.

**Comment by ulysses4ever** [comment]
@Mikolaj you added this issue to the project about replacing `v1-`commands, but I don't quite see which `v1` functionality this would replace. At least, I don't think there's anything better in `v1` than already shown here `--no-code`, which works perfectly with `v2`. The issue is about wrapping it into a prettier interface, but that's unrelated to `v1` I think?

So, how about we remove this from the project?

**Comment by Mikolaj** [comment]
Yes, I guess the workaround is good enough not to make this issue a blocker (though probably not to close the issue and, in any case, that's for the affected users to opine on). Let me remove this ticket from the https://github.com/haskell/cabal/projects/12 project. Thanks.

**Comment by MangoIV** [comment]
I guess there's a similar blocker on why we cannot do `cabal repl all` to have something like `cabal --ghc-options="-fno-code -fwrite-interface"` work on `cabal.project`s with multiple libraries?

**Comment by mpilgrem** [comment]
The Stack project would be interested in Cabal (the library) supporting this. The context is 'nimble workflows', where people seek to have the benefit of type-checking without bearing the 'cost' of compilation. EDIT: Adding for context some discussions at the Haskell Community:
* https://discourse.haskell.org/t/the-fastest-way-to-feed-ghc-type-errors-to-llm/13827
* https://discourse.haskell.org/t/how-to-achieve-anything-like-cabal-typecheck/11214
